subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -164,36 +178,32 @@ public boolean shouldWatch(FileSystemLocationSnapshot snapshot) {
         return !ignoredForWatching(snapshot) && isInWatchableHierarchy(snapshot.getAbsolutePath());
     }
 
-    public void updateUnwatchableFileSystems(WatchMode watchMode) {
-        unwatchableHierarchies = watchUnsupportedFileSystems(watchMode)
+    /**
+     * Detects and updates the unsupported file systems.
+     *
+     * Depending on the watch mode, actually detecting the unsupported file systems may not be necessary.
+     */
+    public void updateUnsupportedFileSystems(WatchMode watchMode) {
+        unsupportedHierarchies = shouldWatchUnsupportedFileSystems(watchMode)
             ? DefaultFileHierarchySet.of()
-            : detectUnsupportedFileSystems();
+            : detectUnsupportedHierarchies();
     }
```

Comment:
I think it would be nicer if the detection happened outside of this class, and only the resulting `FileHierarchySet` would be passed in here. WDYT?

Reply:
I tried it out, and it currently seems that having watchable hierarchies to take care of the unsupported file system detection for
- updating unsupported hierarchies, and
- invalidating unsupported file systems

makes sense. I'd leave it as is right now.

Reply:
OK.